File: simba-core/src/main/java/me/ahoo/simba/core/AbstractMutexContender.java
/*
 * Copyright [2021-2021] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.simba.core;

import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import lombok.extern.slf4j.Slf4j;

/**
 * @author ahoo wang
 */
@Slf4j
public abstract class AbstractMutexContender implements MutexContender {
    private final String mutex;
    private final String contenderId;
    
    public AbstractMutexContender(String mutex) {
        this(mutex, ContenderIdGenerator.Simple.HOST.generate());
    }
    
    public AbstractMutexContender(String mutex, String contenderId) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(mutex), "mutex can not be null or empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(contenderId), "contenderId can not be null or empty!");
        this.mutex = mutex;
        this.contenderId = contenderId;
    }
    
    @Override
    public String getMutex() {
        return mutex;
    }
    
    @Override
    public String getContenderId() {
        return contenderId;
    }
    
    @Override
    public void onAcquired(MutexState mutexState) {
        if (log.isInfoEnabled()) {
            log.info("onAcquired - mutex:[{}] - contenderId:[{}].", mutex, contenderId);
        }
    }
    
    @Override
    public void onReleased(MutexState mutexState) {
        if (log.isInfoEnabled()) {
            log.info("onReleased - mutex:[{}] - contenderId:[{}].", mutex, contenderId);
        }
    }
    
}

File: simba-core/src/main/kotlin/me/ahoo/simba/core/AbstractMutexContender.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.simba.core

import io.github.oshai.kotlinlogging.KotlinLogging

/**
 * Abstract Mutex Contender.
 *
 * @author ahoo wang
 */
abstract class AbstractMutexContender(
    final override val mutex: String,
    final override val contenderId: String = ContenderIdGenerator.HOST.generate()
) : MutexContender {
    companion object {
        private val log = KotlinLogging.logger {}

        /**
         * Aligned with the strictest backend schema (simba_jdbc: `mutex varchar(66)`, `owner_id varchar(128)`).
         */
        const val MAX_MUTEX_LENGTH = 66
        const val MAX_CONTENDER_ID_LENGTH = 128

        /**
         * Owner-event / acquire-result wire delimiter: a contenderId containing it breaks the redis protocol parsing.
         */
        const val CONTENDER_ID_DELIMITER = "@@"
    }

    init {
        require(mutex.isNotBlank()) { "mutex must not be blank!" }
        require(mutex.length <= MAX_MUTEX_LENGTH) { "mutex must not exceed $MAX_MUTEX_LENGTH characters!" }
        require(contenderId.isNotBlank()) { "contenderId must not be blank!" }
        require(contenderId.length <= MAX_CONTENDER_ID_LENGTH) {
            "contenderId must not exceed $MAX_CONTENDER_ID_LENGTH characters!"
        }
        require(!contenderId.contains(CONTENDER_ID_DELIMITER)) {
            "contenderId must not contain [$CONTENDER_ID_DELIMITER]!"
        }
    }

    override fun onAcquired(mutexState: MutexState) {
        log.info {
            "onAcquired - mutex:[$mutex] - contenderId:[$contenderId]."
        }
    }

    override fun onReleased(mutexState: MutexState) {
        log.info {
            "onReleased - mutex:[$mutex] - contenderId:[$contenderId]."
        }
    }
}
